include/pybind11/pytypes.h

```diff
@@ -890,6 +891,22 @@ class iterator : public object {
     object value = {};
 };
 
+
+
+class type : public object {
+public:
+    PYBIND11_OBJECT_COMMON(type, object, PyType_Check)
+
+    explicit type(handle h): type((PyObject*) Py_TYPE(h.ptr()), borrowed_t{}) {}
+    explicit type(object ob): type((PyObject*) Py_TYPE(ob.ptr()), borrowed_t{}) {}
```

Comment:
This isn't necessary, I think? `object` should be a subclass of `handle`, and there's no extra refcounting going on?

Reply:
Oh, yes, some slicing, since it's not by reference. But there are precedents for that, I think?

Reply:
If you/we choose to keep the `object` overload, better to make it a `const object &`, then. It avoids an unnecessary refcount.

Reply:
Ah, no, `PYBIND11_OBJECT` already adds that overload.

https://github.com/pybind/pybind11/blob/cc982ac1cd3ec31040c0cfff3a6dd0183f7c1cac/include/pybind11/pytypes.h#L813-L817

How do we fix this such that it works in the intended way? When do we want `py::type(...)` to "cast" a `py::object` to a subclass, and when do we want it to actually call `type(...)`? Does this mean that @henryiii's original `py::type::of` is necessary, and that we can't have the `py::type(obj)` interface I've been dreaming of? :-(

Reply:
We use `PYBIND11_OBJECT_COMMON` instead of `PYBIND11_OBJECT`.

Reply:
(I'll try dropping this overload soon)

Reply:
Can this be a `const handle &`?

Reply:
> We use `PYBIND11_OBJECT_COMMON` instead of `PYBIND11_OBJECT`.

🤦 How did I miss that?

> Can this be a `const handle &`?

Probably, but I don't think it's necessary. `handle` really is a fancy and more beautiful way of writing `PyObject *`.